sklearn/utils/tests/test_validation.py

```diff
@@ -1961,3 +1964,84 @@ def test_check_array_multiple_extensions(
     X_regular_checked = check_array(X_regular, dtype=None)
     X_extension_checked = check_array(X_extension, dtype=None)
     assert_array_equal(X_regular_checked, X_extension_checked)
+
+
+@pytest.mark.parametrize(
+    "sparse_container",
+    CSR_CONTAINERS
+    + CSC_CONTAINERS
+    + COO_CONTAINERS
+    + DOK_CONTAINERS
+    + DIA_CONTAINERS
+    + LIL_CONTAINERS,
+)
+@pytest.mark.parametrize("output_format", ["csr", "csc", "coo"])
+def test_check_array_downcast_indices_dtype(sparse_container, output_format):
+    """Check the consistency of the indices dtype with sparse matrices/arrays."""
+    X = sparse_container([[0, 1], [1, 0]], dtype=np.float64)
+    X_checked = check_array(X, accept_sparse=output_format)
+    if output_format == "coo":
+        assert X_checked.row.dtype == np.int32
+        assert X_checked.col.dtype == np.int32
+    else:  # output_format in ["csr", "csc"]
+        assert X_checked.indices.dtype == np.int32
+        assert X_checked.indptr.dtype == np.int32
+
+
+def test_smallest_admissible_index_dtype():
+    imax = np.int64(np.iinfo(np.int32).max)
+    too_big = imax + 1
+
+    # Check an empty array
+    a1 = np.ones(0, dtype="uint32")
+    assert np.dtype(
+        _smallest_admissible_index_dtype(a1, check_contents=True)
+    ) == np.dtype("int32")
+
+    # Check with a single array
+    a1 = np.ones(90, dtype="uint32")
+    assert np.dtype(
+        _smallest_admissible_index_dtype(a1, check_contents=True)
+    ) == np.dtype("int32")
+
+    # Check that uint32's with no values too large doesn't return
+    # int64
+    a2 = np.ones(90, dtype="uint32")
+    assert np.dtype(
+        _smallest_admissible_index_dtype((a1, a2), check_contents=True)
+    ) == np.dtype("int32")
+
+    # Check that if we can not convert but all values are less than or
+    # equal to max that we can just convert to int32
+    a1[-1] = imax
+    assert np.dtype(
+        _smallest_admissible_index_dtype((a1, a2), check_contents=True)
+    ) == np.dtype("int32")
+
+    # Check that if it can not convert directly and the contents are
+    # too large that we return int64
+    a1[-1] = too_big
```

Comment:
For instance I would find it more explicit to directly call the `_smallest_admissible_index_dtype` function on `np.full(shape=3, fill_value=np.iinfo(np.int32).max + 1, dtype=np.int64)` or something like that. This way, all the info to understand the test case is present on the line without any indirection to ambiguous variable names.